src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs

```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
+                        }
                     }
+
+                    // Write the bundle manifest
+                    headerOffset = BundleManifest.Write(writer);
+                    _tracer.Log($"Header Offset={headerOffset}");
+                    _tracer.Log($"Meta-data Size={writer.BaseStream.Position - headerOffset}");
+                    _tracer.Log($"Bundle: Path={bundlePath}, Size={bundleStream.Length}");
                 }
 
-                // Write the bundle manifest
-                headerOffset = BundleManifest.Write(writer);
-                _tracer.Log($"Header Offset={headerOffset}");
-                _tracer.Log($"Meta-data Size={writer.BaseStream.Position - headerOffset}");
-                _tracer.Log($"Bundle: Path={bundlePath}, Size={bundle.Length}");
+                var bundleName = Path.GetFileName(bundlePath);
+                SetAsBundle(bundleName, bundleStream, headerOffset, signatureBlob);
+
+                HostWriter.Chmod755(bundlePath);
+
+                return bundlePath;
             }
+        }
 
-            HostWriter.SetAsBundle(bundlePath, headerOffset, _macosCodesign);
+        private void SetAsBundle(string bundleName, FileStream bundleStream, long headerOffset, EmbeddedSignatureBlob? signatureBlob)
+        {
+
+            byte[] bundleHeaderPlaceholder = {
+                    // 8 bytes represent the bundle header-offset
+                    // Zero for non-bundle apphosts (default).
+                    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                    // 32 bytes represent the bundle signature: SHA-256 for ".net core bundle"
+                    0x8b, 0x12, 0x02, 0xb9, 0x6a, 0x61, 0x20, 0x38,
+                    0x72, 0x7b, 0x93, 0x02, 0x14, 0xd7, 0xa0, 0x32,
+                    0x13, 0xf5, 0xb9, 0xe6, 0xef, 0xae, 0x33, 0x18,
+                    0xee, 0x3b, 0x2d, 0xce, 0x24, 0xb3, 0x6a, 0xae
+                };
+            long bundleSize = bundleStream.Length;
+            long signedSize = bundleSize + MachObjectFile.GetSignatureSizeEstimate((uint)bundleSize, bundleName);
+            bundleStream.SetLength(signedSize);
+
+            // Re-write the destination apphost with the proper contents.
+            RetryUtil.RetryOnIOError(() =>
+            {
+                using (MemoryMappedFile bundleMap = MemoryMappedFile.CreateFromFile(bundleStream, null, bundleStream.Length, MemoryMappedFileAccess.ReadWrite, HandleInheritability.None, leaveOpen: true))
+                using (MemoryMappedViewAccessor accessor = bundleMap.CreateViewAccessor(0, 0, MemoryMappedFileAccess.ReadWrite))
+                {
+                    BinaryUtils.SearchAndReplace(accessor,
+                                                bundleHeaderPlaceholder,
+                                                BitConverter.GetBytes(headerOffset),
+                                                pad0s: false);
+
+                    if (_target.IsOSX)
```

Comment:
Seems strange to me that this bit and the codesign header removal are in separate places? Shouldn't they be in the same place?